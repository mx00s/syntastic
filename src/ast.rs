//! Abstract syntax trees
//!
//! [Ast]s are trees with [Variable]s, `V`, as leaves and operations as non-leaves.
//! Each [Ast] and all of its subcomponents are indexed by some syntactic sort, `S`.
//! Each [Operator], `O`, encodes a sequence of sorts and can be composed with [Ast]s of those
//! sorts to construct an operation node.

#[cfg(test)]
mod tests;

mod types;
pub use self::types::*;

#[cfg(feature = "with-proptest")]
mod with_proptest;
#[cfg(feature = "with-proptest")]
pub use self::with_proptest::*;

#[cfg(test)]
mod bidirectional;
