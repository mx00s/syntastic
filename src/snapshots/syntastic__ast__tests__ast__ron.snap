---
source: src/ast.rs
expression: example_ast()
---
Operation(
  operator: Operator(
    op: Plus,
    sort: Num,
    arity: [
      Num,
      Num,
    ],
  ),
  args: [
    Value(
      val: 2,
      sort: Num,
    ),
    Operation(
      operator: Operator(
        op: Times,
        sort: Num,
        arity: [
          Num,
          Num,
        ],
      ),
      args: [
        Value(
          val: 3,
          sort: Num,
        ),
        Variable(
          var: X,
          sort: Num,
        ),
      ],
    ),
  ],
)
